File: hs/js1/r3-2.hs
{- cabal:
build-depends: base
-}

import System.Exit (die)
import System.IO (hFlush, stdout)
import Text.Read (readMaybe)

putStrFlush :: String -> IO ()
putStrFlush s = do
    putStr s
    hFlush stdout

getFloat :: IO Float
getFloat = do
    line <- getLine
    case readMaybe line of
        Just a -> return a
        Nothing -> do
            die "Not a number"

showAnswer :: Float -> Float -> Float -> String
showAnswer 0 b a2 = "x1 = " ++ show (-b / a2)
showAnswer d b a2
    | d < 0 = "Not a quadratic equation"
    | otherwise =
        let sqrt_d = sqrt d
            x1 = (-b + sqrt_d) / a2
            x2 = (-b - sqrt_d) / a2
         in "x1 = " ++ show x1 ++ "\nx2 = " ++ show x2

main :: IO ()
main = do
    putStrFlush "a = "
    a <- getFloat
    putStrFlush "b = "
    b <- getFloat
    putStrFlush "c = "
    c <- getFloat

    let d = b * b - 4 * a * c
        double_a = 2 * a
     in putStrLn $ showAnswer d b double_a

File: second_grade/js1/k3/r3-2.c
/* 23 豊島圭吾
 * 第3回レポート プログラム2
 */

#include <math.h>
#include <stdio.h>

int main(void) {
  double a, b, c;
  printf("a = ");
  scanf("%lf", &a);
  printf("b = ");
  scanf("%lf", &b);
  printf("c = ");
  scanf("%lf", &c);

  double d = b * b - 4 * a * c;
  double double_a = 2 * a, sqrt_d = sqrt(d);
  if (d < 0) {
    printf("解なし\n");
  } else if (d == 0) {
    printf("x = %.2lf\n", (-b + sqrt_d) / double_a);
  } else {
    printf("x1 = %.2lf\n", (-b + sqrt_d) / double_a);
    printf("x2 = %.2lf\n", (-b - sqrt_d) / double_a);
  }

  return 0;
}

/* 実行結果
a = 1
b = 2
c = 1
x = -1.00

a = 1
b = 1
c = 1
解なし
 */
